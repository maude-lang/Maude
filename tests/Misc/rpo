#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/rpo.maude -no-banner -no-advise \
  > rpo.out 2>&1

diff $srcdir/rpo.expected rpo.out > /dev/null 2>&1
