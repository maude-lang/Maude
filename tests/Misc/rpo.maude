***
***	Test basic use of the rpo attribute
***

set show timing off .

*** don't pull in BOOL so we have simple modules to look at
set include BOOL off .


*** The rpo attribute appears when printing the module

fmod M1 is
  sort Bar .
  op a : Bar -> Bar [rpo 1] .
  op b : Bar -> Universal [poly (0) rpo 2] .
endfm

show module .
show ops .

*** The rpo attribute is copied when the module is imported

fmod M2 is
  including M1 .
  sort Foo .
  op c : Foo -> Foo [rpo 3] .
endfm

show ops .

fmod M3 is
  including M1 * (op a to a [rpo 4], op b to b [rpo 5]) .
endfm

show module .
show ops .

*** Conflicts are detected

fth C1 is
  sorts Bar .
  op a : Bar -> Bar [rpo 1 rpo 2] .
  op b : Bar -> Bar [rpo] .
  op b : Bar -> Bar [rpo bad] .
endfth

fth C2 is
  sorts Bar Foo .
  subsort Foo < Bar .
  op a : Bar -> Bar [rpo 1] .
  op a : Foo -> Foo [rpo 2] .
endfth

show module .

*** The attribute appears at the metalevel

reduce in META-LEVEL : upOpDecls('M1, false) .
reduce upOpDecls('M2, true) .
reduce upModule('M3, false) .
reduce upOpDecls('M3, true) .
