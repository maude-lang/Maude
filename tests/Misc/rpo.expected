fmod M1 is
  sort Bar .
  op a : Bar -> Bar [rpo 1] .
  op b : Bar -> Universal [poly (0) rpo 2] .
endfm
op b : Bar -> Universal [poly (0) rpo 2] .
op a : Bar -> Bar [rpo 1] .
op b : Bar -> Universal [poly (0) rpo 2] .
op a : Bar -> Bar [rpo 1] .
op c : Foo -> Foo [rpo 3] .
fmod M3 is
  including M1 * (op a to a [rpo 4], op b to b [rpo 5]) .
endfm
op b : Bar -> Universal [poly (0) rpo 5] .
op a : Bar -> Bar [rpo 4] .
Warning: <standard input>, line 43 (fth C1): multiple rpo attributes.
Warning: <standard input>, line 44 (fth C1): rpo attribute without value in
    operator declaration.
Warning: <standard input>, line 45 (fth C1): bad value bad for rpo attribute.
fth C2 is
  sorts Bar Foo .
  subsort Foo < Bar .
  op a : Bar -> Bar [rpo 1] .
  op a : Foo -> Foo [rpo 2] .
endfth
==========================================
reduce in META-LEVEL : upOpDecls('M1, false) .
rewrites: 1
result OpDeclSet: op 'a : 'Bar -> 'Bar [rpo(1)] .
op 'b : 'Bar -> 'Universal [rpo(2) poly(0)] .
==========================================
reduce in META-LEVEL : upOpDecls('M2, true) .
rewrites: 1
result OpDeclSet: op 'a : 'Bar -> 'Bar [rpo(1)] .
op 'b : 'Bar -> 'Universal [rpo(2) poly(0)] .
op 'c : 'Foo -> 'Foo [rpo(3)] .
==========================================
reduce in META-LEVEL : upModule('M3, false) .
rewrites: 1
result FModule: fmod 'M3 is
  including 'M1 * (
    op 'a to 'a [rpo(4)],
    op 'b to 'b [rpo(5)]
  ) .
  sorts none .
  none
  none
  none
  none
endfm
==========================================
reduce in META-LEVEL : upOpDecls('M3, true) .
rewrites: 1
result OpDeclSet: op 'a : 'Bar -> 'Bar [rpo(4)] .
op 'b : 'Bar -> 'Universal [rpo(5) poly(0)] .
Bye.
